core/src/com/unciv/logic/city/CityInfo.kt

```diff
@@ -175,60 +181,105 @@ class CityInfo {
-            civInfo.gameInfo.civilizations.flatMap { it.cities.map { city -> city.name } }
-        // We take the last unused city name for each other civ in this game, skipping civs whose
-        // names are exhausted, and choose a random one from that pool if it's not empty.
-        var newNames = civInfo.gameInfo.civilizations
-            .filter { it.isMajorCiv() && it != civInfo }
-            .mapNotNull {
-                it.nation.cities
-                    .lastOrNull { city -> city !in usedCityNames }
-            }
-        if (newNames.isNotEmpty()) {
-            return newNames.random()
-        }
+    /**
+     * Borrows a city name from another major civilization.
+     *
+     * @param foundingCiv The civilization that founded this city.
+     * @param aliveCivs Every civilization currently alive.
+     * @param usedCityNames Every city name that have already been taken.
+     * @return A new city named in [String]. Null if failed to generate a name.
+     */
+    private fun borrowCityName(
+        foundingCiv: CivilizationInfo,
+        aliveCivs: Set<CivilizationInfo>,
+        usedCityNames: Set<String>
+    ): String? {
+        val aliveMajorNations: Sequence<Nation> =
+            aliveCivs.asSequence().filter { civ -> civ.isMajorCiv() }.map { civ -> civ.nation }
+
+        /*
+        We take the last unused city name for each other major nation in this game,
+        skipping nations whose names are exhausted,
+        and choose a random one from that pool if it's not empty.
+        */
+        val otherMajorNations: Sequence<Nation> =
+            aliveMajorNations.filter { nation -> nation != foundingCiv.nation }
+        var newCityNames: Set<String> =
+            otherMajorNations.mapNotNull { nation ->
+                nation.cities.lastOrNull { city -> city !in usedCityNames }
+            }.toSet()
+        if (newCityNames.isNotEmpty()) return newCityNames.random()
 
         // As per fandom wiki, once the names from the other nations in the game are exhausted,
-        // names are taken from the rest of the nations in the ruleset
-        newNames = getRuleset()
-            .nations
-            .filter { it.key !in civInfo.gameInfo.civilizations.map { civ -> civ.nation.name } }
-            .values
-            .map {
-                it.cities
-                    .filter { city -> city !in usedCityNames }
-            }.flatten()
-        if (newNames.isNotEmpty()) {
-            return newNames.random()
-        }
+        // names are taken from the rest of the major nations in the rule set
+        val absentMajorNations: Sequence<Nation> =
+            getRuleset().nations.values.asSequence().filter { nation ->
+                nation.isMajorCiv() && nation !in aliveMajorNations
+            }
+        newCityNames =
+            absentMajorNations.flatMap { nation ->
+                nation.cities.asSequence().filter { city -> city !in usedCityNames }
+            }.toSet()
```

Comment:
If you're changing the type anyways, why not a hashset?

Reply:
My take is - Set is a much more general contract than HashSet. With toSet you leave some choice to the compiler, which is good. And lookup on a generic Set can't be less performant than on a HashSet, or else JetBrains would have none. Look under the hood - it's implemented as a LinkedHashSet with optimizeReadOnlySet called on it. Next version might implement it differently if they find some room for improvement...

So, unless you know you need a specific Set implementation, toSet ***is*** best practice.

Reply:
My original thought process here was 'I need an immutable or read-only set'. Also, aren't LinkedHashSets slightly faster than normal HashSets? Then again HashSets consume less memory than the other...so maybe I should indeed convert the result to HashSet.

Reply:
I've eventually decided to leave it as `toSet()`. After all it's just some sets of strings and this method is called once in every 5 minutes or so, which is not much considering there are methods and functions called every single game loop. I'm gonna believe in the compiler.